server/src/main/java/io/crate/planner/statement/CopyFromPlan.java

```diff
@@ -228,29 +232,42 @@ public static ExecutionPlan planCopyFromExecution(AnalyzedCopyFrom copyFrom,
             .map(table::getReference)
             .collect(Collectors.toList());
 
-        List<Symbol> toCollect = getSymbolsRequiredForShardIdCalc(
+        // Those columns must be collected regardless of their presence in targetColumns
+        // since they are used for calculating shard id.
+        // Collecting all top-level table columns might miss those columns if they are sub-columns, hence special collecting.
+        Set<Reference> targetColumns = getColumnsRequiredForShardIdCalc(
             primaryKeyRefs,
             table.partitionedByColumns(),
             clusteredBy == null ? null : table.getReference(clusteredBy)
         );
-        Reference rawOrDoc = rawOrDoc(table, partitionIdent);
-        final int rawOrDocIdx = toCollect.size();
-        toCollect.add(rawOrDoc);
 
-        String[] excludes = partitionedByNames.size() > 0
-            ? partitionedByNames.toArray(new String[0]) : null;
+        List<Symbol> toCollect = new ArrayList<>(targetColumns);
 
-        InputColumns.SourceSymbols sourceSymbols = new InputColumns.SourceSymbols(toCollect);
+        // InputColumns.create(generatedReference, sourceSymbols) would reset COALESCE usage to original expression.
+        // This map is used for direct InputColumn creation out of concrete COALESCE function and also for setting partition literals.
+        Map<ColumnIdent, io.crate.expression.symbol.Function> coalesceFunctionsByGenRef = new HashMap<>();
+
+        // Initialized when all refs are added.
+        // In summary mode we add RAW (for now), in regular mode we add not-yet added top-level target columns.
+        InputColumns.SourceSymbols sourceSymbols;
         Symbol clusteredByInputCol = null;
-        if (clusteredBy != null) {
-            clusteredByInputCol = InputColumns.create(table.getReference(clusteredBy), sourceSymbols);
-        }
 
-        SourceIndexWriterProjection sourceIndexWriterProjection;
+        AbstractIndexWriterProjection indexWriterProjection;
         List<? extends Symbol> projectionOutputs = AbstractIndexWriterProjection.OUTPUTS;
         boolean returnSummary = copyFrom instanceof AnalyzedCopyFromReturnSummary;
         boolean failFast = boundedCopyFrom.settings().getAsBoolean("fail_fast", false);
         if (returnSummary || failFast) {
+
+            Reference rawOrDoc = rawOrDoc(table, partitionIdent);
+            final int rawOrDocIdx = toCollect.size();
+            toCollect.add(rawOrDoc);
+
+            sourceSymbols = new InputColumns.SourceSymbols(toCollect);
+
+            if (clusteredBy != null) {
+                clusteredByInputCol = InputColumns.create(table.getReference(clusteredBy), sourceSymbols);
+            }
```

Comment:
RAW will be gone once I migrate RETURN SUMMARY, kept it for now in order to mute fewer tests related to summary.